akka-stream-tests/src/test/scala/akka/stream/scaladsl/HubSpec.scala

```diff
@@ -367,4 +367,231 @@ class HubSpec extends StreamSpec {
 
   }
 
+  "PartitionHub" must {
+
+    "work in the happy case with one stream" in assertAllStagesStopped {
+      val source = Source(1 to 10).runWith(PartitionHub.sink(() ⇒ (ids, elem) ⇒ ids(0), startAfterNbrOfStreams = 0, bufferSize = 8))
+      source.runWith(Sink.seq).futureValue should ===(1 to 10)
+    }
+
+    "work in the happy case with two streams" in assertAllStagesStopped {
+      val source = Source(0 until 10).runWith(PartitionHub.sink(() ⇒ (ids, elem) ⇒ ids(elem % ids.length), startAfterNbrOfStreams = 2, bufferSize = 8))
+      val result1 = source.runWith(Sink.seq)
+      // it should not start publishing until startAfterNbrOfStreams = 2
+      Thread.sleep(20)
+      val result2 = source.runWith(Sink.seq)
+      result1.futureValue should ===(0 to 8 by 2)
+      result2.futureValue should ===(1 to 9 by 2)
+    }
+
+    "be able to use as round-robin router" in assertAllStagesStopped {
+      val source = Source(0 until 10).runWith(PartitionHub.sink(() ⇒ {
+        var n = 0L
+
+        (ids, elem) ⇒ {
+          n += 1
+          ids((n % ids.length).toInt)
+        }
+      }, startAfterNbrOfStreams = 2, bufferSize = 8))
+      val result1 = source.runWith(Sink.seq)
+      val result2 = source.runWith(Sink.seq)
+      result1.futureValue should ===(1 to 9 by 2)
+      result2.futureValue should ===(0 to 8 by 2)
+    }
+
+    "be able to use as sticky session router" in assertAllStagesStopped {
+      val source = Source(List("usr-1", "usr-2", "usr-1", "usr-3")).runWith(PartitionHub.sink(() ⇒ {
+        var sessions = Map.empty[String, Long]
+        var n = 0L
+
+        (ids, elem) ⇒ {
+          sessions.get(elem) match {
+            case Some(id) if ids.exists(_ == id) ⇒ id
+            case _ ⇒
+              n += 1
+              val id = ids((n % ids.length).toInt)
+              sessions = sessions.updated(elem, id)
+              id
+          }
+        }
+      }, startAfterNbrOfStreams = 2, bufferSize = 8))
+      val result1 = source.runWith(Sink.seq)
+      val result2 = source.runWith(Sink.seq)
+      result1.futureValue should ===(List("usr-2"))
+      result2.futureValue should ===(List("usr-1", "usr-1", "usr-3"))
+    }
```

Comment:
I have added the more powerful API: `() => (Array[Long], T) => Long` and above tests illustrate how that can be used to implement stateful routing. Question is if we also need the more simple index based API `(Int, T) => Int`?

The happy case above would then be:
```
PartitionHub.simpleSink((size, elem) ⇒ elem % size)
```

or should we make it even more specialized for `hashCode` like scenarios: `T => Int`, e.g.
```
PartitionHub.hashingSink(elem ⇒ elem.hashCode)
```